proselint/checks/__init__.py

```diff
@@ -1 +1,394 @@
+    """Extract the checks.
+    Rule: fn-name must begin with "check", so check_xyz() is ok
+    """
+    # TODO: benchmark consecutive runs of this
+    #       config should only translate once to check-list
+    checks = []
+    check_names = [key for (key, val) in options["checks"].items() if val]
+
+    for check_name in check_names:
+        try:
+            module = importlib.import_module("." + check_name, "proselint.checks")
+        except ModuleNotFoundError:
+            log.exception(
+                "requested config-flag '%s' not found in proselint.checks",
+                check_name,
+            )
+            continue
+        checks += [getattr(module, d) for d in dir(module) if re.match(r"^check", d)]
+
+    log.debug("Collected %d checks to run", len(checks))
+    return checks
+
+
+def run_checks(_check: Callable, _text: str, source: str = "") -> list:
+    # TODO: frozenset is hashable (list without duplicates) -> check-list, result-lists
+    # padding
+    # -> some checks expect words in text and need something around it
+    # -> this prevents edge-cases
+    _text = f" \n{_text}\n "  # maybe not the fastest OP
+    errors = []
+    results = _check(_text)
+    for result in results:
+        (start, end, check_name, message, replacements) = result
+        (line, column) = get_line_and_column(_text, start)
+        if not is_quoted(start, _text):
+            # note:
+            #    - switch to 1based counting by adding +1
+            #    - for line it cancels out with -1 from padding
+            errors += [
+                (
+                    check_name,
+                    message,
+                    source,  # can't be Path, unless pickle changes
+                    line,  # +1 -1, cancel out
+                    column + 1,
+                    start + 1,
+                    end + 1,
+                    end - start,
+                    "warning",
+                    replacements,
+                ),
+            ]
+    return errors
+
+
+def get_line_and_column(text, position):
+    """Return the line number and column of a position in a string."""
+    position_counter = 0
+    line_no = 0
+    for line in text.splitlines(True):
+        if (position_counter + len(line.rstrip())) >= position:
+            break
+        position_counter += len(line)
+        line_no += 1
+    return line_no, position - position_counter
+
+
+def is_quoted(position: int, text: str) -> bool:
+    """Determine if the position in the text falls within a quote."""
+
+    def matching(quotemark1: str, quotemark2: str) -> bool:
+        straight = "\"'"
+        curly = "“”"
+        return (quotemark1 in straight and quotemark2 in straight) or (
+            quotemark1 in curly and quotemark2 in curly
+        )
+
+    def find_ranges(_text: str) -> list[tuple[int, int]]:
+        # TODO: optimize, as it is top3 time-waster (of module-functions)
+        #       this could be a 1d array / LUT
```

Comment:
```suggestion
        # FIXME: optimize - this function is one of the 3 most expensive in the module
        # this could be a 1-dimensional array or lookup table
```

Reply:
done